watertap/costing/watertap_costing_package.py

```diff
@@ -859,6 +914,109 @@ def cost_crystallizer_by_volume(blk):
 
+    @staticmethod
+    def cost_gac(blk):
+        """
+        3 equation capital cost estimation for GAC
+            contactor/pressure vessel cost by polynomial with single contactor volume
+            initial charge of GAC adsorbent cost by exponential with GAC mass
+            other process costs calculated power law with total contactor volume
+        """
+        make_capital_cost_var(blk)
+        blk.contactor_cost = pyo.Var(
+            initialize=1e5,
+            domain=pyo.NonNegativeReals,
+            units=blk.costing_package.base_currency,
+            doc="Unit contactor capital cost",
+        )
+        blk.adsorbent_unit_cost = pyo.Var(
+            initialize=2,
+            domain=pyo.NonNegativeReals,
+            units=blk.costing_package.base_currency * pyo.units.kg**-1,
+            doc="Unit adsorbent unit cost per mass",
+        )
+        blk.adsorbent_cost = pyo.Var(
+            initialize=1e5,
+            domain=pyo.NonNegativeReals,
+            units=blk.costing_package.base_currency,
+            doc="Unit adsorbent capital cost",
+        )
+        blk.other_process_cost = pyo.Var(
+            initialize=1e5,
+            domain=pyo.NonNegativeReals,
+            units=blk.costing_package.base_currency,
+            doc="Unit other process capital cost",
+        )
+
+        # TODO: currently costed as 1 operating, 1 redundant vessel (gac_num_contactor = 2)
+        #  where flexibility can be added to volume by number in operation and add redundant vessels
+        blk.contactor_cost_constraint = pyo.Constraint(
+            expr=blk.contactor_cost
+            == blk.costing_package.gac_num_contactors
+            * pyo.units.convert(
+                (
+                    blk.costing_package.gac_contactor_cost_coeff_3
+                    * blk.unit_model.bed_volume**3
+                    + blk.costing_package.gac_contactor_cost_coeff_2
+                    * blk.unit_model.bed_volume**2
+                    + blk.costing_package.gac_contactor_cost_coeff_1
+                    * blk.unit_model.bed_volume**1
+                    + blk.costing_package.gac_contactor_cost_coeff_0
+                ),
+                to_units=blk.costing_package.base_currency,
+            )
+        )
+        if blk.unit_model.bed_mass_gac.value < 18143.68:
+            blk.adsorbent_unit_cost_constraint = pyo.Constraint(
+                expr=blk.adsorbent_unit_cost
+                == pyo.units.convert(
+                    blk.costing_package.gac_adsorbent_unit_cost_coeff
+                    * pyo.exp(
+                        blk.unit_model.bed_mass_gac
+                        * blk.costing_package.gac_adsorbent_unit_cost_exp_coeff
+                    ),
+                    to_units=blk.costing_package.base_currency * pyo.units.kg**-1,
+                )
+            )
+        else:
+            blk.adsorbent_unit_cost_constraint = pyo.Constraint(
+                expr=blk.adsorbent_unit_cost

```

Comment:
Just a thought- do we want to consider using an Expr_IF here?

Reply:
Also, you'll want a test to make sure this works as intended.

Reply:
I agree that there is a better way to write this. This is simply trying to impose a maximum on the variable that if purchasing more than the 18143.68 kg, the unit price isn't improved. Didn't know whether to overwrite a temp variable and write only one equation or do something similar to this, but I would like to find a current WaterTAP model that handles this kind of if statement to recode in a similar way.

Reply:
I'm assuming Expr_IF handles conditionals flipping between true and false as a variable moves from the initial to solution value? That would be the concern with how it's currently done.

Reply:
I would consider using `smooth_min`, e.g., `bed_mass_for_cost == smooth_min(18143.68, blk.unit_model.bed_mass_gac)`

Reply:
Either an `expr_if` or a `smooth_min` would work here, and both have advantages and disadvantages.

`smooth_min` is smooth and continuous which is good, but it is not the nicest non-linear constraint.
`expr_if` on the other hand does have an issue with being discontinuous in the 1st derivative, which can cause issues if the solution lies close to that point, but is otherwise probably OK.